rfcs/20191212-keras-categorical-inputs.md

```diff
@@ -0,0 +1,401 @@
+* Introduce sparse inputs that work with Keras linear models and other layers that support sparsity
+
+Other proposed layers for replacement of feature columns such as `tf.feature_column.bucketized_column` and `tf.feature_column.numeric_column` has been discussed [here](https://github.com/keras-team/governance/blob/master/rfcs/20190502-preprocessing-layers.md) and are not the focus of this document.
+
+The proposed layers should support ragged tensors.
+
+## Motivation
+
+Specifically, by introducing the 4 layers, we aim to address these pain points:
+* Users have to define both feature columns and Keras Inputs for the model, resulting in code duplication and deviation from DRY (Do not repeat yourself) principle. See this [Github issue](https://github.com/tensorflow/tensorflow/issues/27416).
+* Users with large dimension categorical inputs will incur large memory footprint and computation cost, if wrapped with indicator column through `tf.keras.layers.DenseFeatures`.
+* Currently there is no way to correctly feed Keras linear model or dense layer with multivalent categorical inputs or weighted categorical inputs.
+
+## User Benefit
+
+We expect to get rid of the user painpoints once migrating off feature columns.
+
+## Example Workflows
+
+Two example workflows are presented below. These workflows can be found at this [colab](https://colab.sandbox.google.com/drive/1cEJhSYLcc2MKH7itwcDvue4PfvrLN-OR#scrollTo=22sa0D19kxXY).
+
+### Workflow 1
+
+The first example gives an equivalent code snippet to canned `LinearEstimator` [tutorial](https://www.tensorflow.org/tutorials/estimator/linear) on the Titanic dataset:
+
+```python
+dftrain = pd.read_csv('https://storage.googleapis.com/tf-datasets/titanic/train.csv')
+y_train = dftrain.pop('survived')
+
+CATEGORICAL_COLUMNS = ['sex', 'n_siblings_spouses', 'parch', 'class', 'deck', 'embark_town', 'alone']
+NUMERICAL_COLUMNS = ['age', 'fare']
+# input list to create functional model.
+model_inputs = []
+# input list to feed linear model.
+linear_inputs = []
+for feature_name in CATEGORICAL_COLUMNS:
+	feature_input = tf.keras.Input(shape=(1,), dtype=tf.string, name=feature_name, sparse=True)
+	vocab_list = sorted(dftrain[feature_name].unique())
+	# Map string values to indices
+	x = tf.keras.layers.Lookup(vocabulary=vocab_list, name=feature_name)(feature_input)
+  x = tf.keras.layers.Vectorize(num_categories=len(vocab_list))(x)
+	linear_inputs.append(x)
+	model_inputs.append(feature_input)
+
+for feature_name in NUMERICAL_COLUMNS:
+	feature_input = tf.keras.Input(shape=(1,), name=feature_name)
+	linear_inputs.append(feature_input)
+	model_inputs.append(feature_input)
+
+linear_model = tf.keras.experimental.LinearModel(units=1)
+linear_logits = linear_model(linear_inputs)
+model = tf.keras.Model(model_inputs, linear_logits)
+
+model.compile('sgd', loss=tf.keras.losses.BinaryCrossEntropy(from_logits=True), metrics=['accuracy'])
+
+dataset = tf.data.Dataset.from_tensor_slices((
```

Comment:
Is it possible to have a single input node and use some layer to decompose the input to these single column nodes?

Reply:
use tf.split for that?